File: airesources/OCaml/networking.ml
open HaliteImpl

let get_init () =
  let (num_players, player_id) = Scanf.scanf "%d %d\n" (fun x y -> (x, Player.Player y)) in
  let players : (Player.id, location) Hashtbl.t = Hashtbl.create num_players in
  for _ = 1 to num_players do
    Scanf.scanf "%d %d %d\n" (fun player x y -> Hashtbl.add players (Player.Player player) {x; y})
  done;
  let (width, height) = Scanf.scanf "%d %d\n" (fun x y -> (x, y)) in
  let cells =
    Array.init height (fun y ->
        Array.init width (fun x ->
            Scanf.scanf "%d%c" (fun energy _ ->
                if x = width - 1 then Scanf.scanf "\n" ();
                ({location={x; y}; energy; entity=None; owner=None; dropoff=None} : Cell.t)
              )
          )
      ) in
  Board.board := Some {width; height; cells};
  Hashtbl.iter (fun id {x; y} ->
      let player : Player.t = {id; factory_cell=cells.(y).(x); dropoffs=[]; energy=0; entities=[]} in
      Hashtbl.add Player.players id player;
      Player.all_players := player::!Player.all_players;
    ) players;
  Player.current_player := Some (Hashtbl.find Player.players player_id)

let get_frame () =
  let turn_number = Scanf.scanf "%d\n" (fun x -> x) in
  let players = Player.players in
  for _ = 1 to Hashtbl.length players do
    Scanf.scanf "%d %d %d %d\n" begin
      fun player_id num_entities num_dropoffs energy ->
        let owner = Hashtbl.find players (Player.Player player_id) in
        owner.energy <- energy;
        owner.entities <- List.init num_entities (fun _ ->
            Scanf.scanf "%d %d %d %d\n" begin
              fun entity x y energy ->
                let id = Entity.Entity entity in
                let cell = Board.at {x; y} in
                match Hashtbl.find_opt Entity.entities id with
                | Some entity ->
                  entity.cell <- cell;
                  entity.energy <- energy;
                  entity
                | None ->
                  let entity : Entity.t = {id; energy; owner; cell} in
                  Hashtbl.add Entity.entities id entity;
                  entity
            end
          );
        owner.dropoffs <- List.init num_dropoffs (fun _ ->
            Scanf.scanf "%d %d %d\n" begin
              fun dropoff x y ->
                let id = Dropoff.Dropoff dropoff in
                match Hashtbl.find_opt Dropoff.dropoffs id with
                | Some dropoff -> dropoff
                | None ->
                  let dropoff : Dropoff.t = {id; owner; cell=Board.at {x; y}} in
                  Hashtbl.add Dropoff.dropoffs id dropoff;
                  dropoff
            end
          )
    end
  done;
  Scanf.scanf "%d\n" begin
    fun num_changed_cells ->
      for _ = 1 to num_changed_cells do
        Scanf.scanf "%d %d %d\n" begin
          fun x y energy -> (Board.at {x; y}).energy <- energy
        end
      done
  end;
  turn_number

let end_turn commands =
  let open Halite in
  let id entity = match Entity.id entity with Entity.Entity id -> id in
  List.iter begin
    function
    | Move (entity, dir) ->
      Printf.printf "m %d %s "
        (id entity)
        (match dir with North -> "n" | South -> "s" | East -> "e" | West -> "w")
    | Spawn energy -> Printf.printf "g %d " energy
    | Dump (entity, energy) -> Printf.printf "d %d %d " (id entity) energy
    | Construct entity -> Printf.printf "c %d " (id entity)
  end commands;
  Printf.printf "\n"; flush stdout

let send_init name = Printf.printf "%s\n" name; flush stdout

File: airesources/Java/hlt/Networking.java
package hlt;

import java.util.Map;
import java.util.ArrayList;
import java.io.FileWriter;
import java.io.IOException;

public class Networking {
    private static Player[] players;
    private static GameMap map;
    private static String moveBuffer = "";
    private static int myID;
    private static int turnNumber;

    private static String readLine() {
        try {
            StringBuilder builder = new StringBuilder();
            int buffer;

            for (; (buffer = System.in.read()) >= 0;) {
                if (buffer == '\n') {
                    break;
                }
                if (buffer == '\r') { // Ignore carriage return if on windows for manual testing.
                    continue;
                }
                builder = builder.append((char)buffer);
            }
            return builder.toString();
        } catch(final Exception e) {
            System.exit(1);
            throw new RuntimeException(e);
        }
    }

    private static String[] getSplitLine() { return readLine().split(" "); }

    public static void getInit() {
        try {
            Log.initialize(new FileWriter("Log.log"));
        } catch (IOException e) {
            e.printStackTrace();
        }
        String[] c = getSplitLine();
        int numPlayers = Integer.parseInt(c[0]);
        myID = Integer.parseInt(c[1]);
        players = new Player[numPlayers];
        for(int i = 0; i < numPlayers; i++) {
            c = getSplitLine();
            Location fl = new Location(Integer.parseInt(c[1]), Integer.parseInt(c[2]));
            players[i] = new Player(Integer.parseInt(c[0]), 0, fl, new ArrayList<Ship>(), new ArrayList<Dropoff>());
        }

        c = getSplitLine();
        int width = Integer.parseInt(c[0]);
        int height = Integer.parseInt(c[1]);
        int[][] grid = new int[height][width];
        for(int y=0; y<height; y++) {
            c = getSplitLine();
            for(int x=0; x<width; x++) {
                grid[y][x] = Integer.parseInt(c[x]);
            }
        }
        map = new GameMap(width, height, grid);
    }


    public static void sendInit(String name) {
        if(name.length() < 1) System.out.println(' ');
        else System.out.println(name);
        System.out.flush();
    }
    public static void getFrame() {
        // TODO give this out
        turnNumber = Integer.parseInt(readLine());
        String[] c;
        for(int i=0; i<players.length; i++) {
            c = getSplitLine();
            int playerID = Integer.parseInt(c[0]);
            int numShips = Integer.parseInt(c[1]);
            int numDropoffs = Integer.parseInt(c[2]);
            int halite = Integer.parseInt(c[3]);
            ArrayList<Ship> ships = new ArrayList<Ship>();
            ArrayList<Dropoff> dropoffs = new ArrayList<Dropoff>();
            for(int j=0; j<numShips; j++) {
                c = getSplitLine();
                ships.add(new Ship(
                    Integer.parseInt(c[0]),
                    new Location(Integer.parseInt(c[1]), Integer.parseInt(c[2])),
                    Integer.parseInt(c[3])));
            }
            for(int j=0; j<numDropoffs; j++) {
                c = getSplitLine();
                dropoffs.add(new Dropoff(
                    Integer.parseInt(c[0]),
                    new Location(Integer.parseInt(c[1]), Integer.parseInt(c[2]))));
            }
            players[playerID] = new Player(playerID,
                halite,
                players[i].getShipyardLocation(),
                ships,
                dropoffs);
        }
        int numChangedCells = Integer.parseInt(readLine());
        for(int i=0; i<numChangedCells; i++) {
            c = getSplitLine();
            map.setSquare(Integer.parseInt(c[0]), Integer.parseInt(c[1]), Integer.parseInt(c[2]));
        }
    }
    public static void sendFrame() {
        System.out.println(moveBuffer);
        System.out.flush();
        moveBuffer = "";
    }

    public static void move(int shipID, Direction direction) {
        if(direction != Direction.STILL) moveBuffer += "m " + shipID + " " + direction.getCharRepresent() + " ";
    }
    public static void dump(int shipID, int haliteAmount) {
        moveBuffer += "d " + shipID + " " + haliteAmount + " ";
    }
    public static void construct(int shipID) {
        moveBuffer += "c " + shipID + " ";
    }
    public static void spawn(int halite) {
        moveBuffer += "g " + halite + " ";
    }
    public static int getTurnNumber() { return turnNumber; }
    public static int getMyID() { return myID; }
    public static GameMap getMap() { return map; }
    public static Player[] getPlayers() { return players; }
}
